Firestore/core/src/firebase/firestore/util/async_queue.cc

```diff
@@ -64,13 +68,40 @@ void AsyncQueue::Enqueue(const Operation& operation) {
   EnqueueRelaxed(operation);
 }
 
+void AsyncQueue::EnqueueAndInitializeShutdown(const Operation& operation) {
+  std::lock_guard<std::mutex> lock{shut_down_mutex_};
+  VerifySequentialOrder();
+  if (is_shutting_down_) {
+    return;
+  }
+  executor_->Execute(Wrap(operation));
+  is_shutting_down_ = true;
+}
+
+void AsyncQueue::EnqueueEvenAfterShutdown(const Operation& operation) {
+  // Still guarding the lock to ensure sequential scheduling.
+  std::lock_guard<std::mutex> lock{shut_down_mutex_};
+  VerifySequentialOrder();
+  executor_->Execute(Wrap(operation));
+}
+
+bool AsyncQueue::is_shutting_down() const {
+  std::lock_guard<std::mutex> lock{shut_down_mutex_};
+  return is_shutting_down_;
+}
+
 void AsyncQueue::EnqueueRelaxed(const Operation& operation) {
+  std::lock_guard<std::mutex> lock{shut_down_mutex_};
```

Comment:
@wilhuff This change seems reasonable in the context of this PR, but I can't help but feel it's unfortunate that the queue now has to lock on every operation, even though shutdown is a very infrequent event (probably not used at all by many users). I don't have a great alternative suggestion off the top of my head, though. What's your take on this?

Reply:
I don't think there's any better alternative. Uncontested lock acquisitions are pretty cheap anyway so this doesn't really worry me.